src/dump.rs

```diff
@@ -0,0 +1,281 @@
+        }
+    }
+}
+
+const DEFAULT_LEAF: CpuIdResult = CpuIdResult {
+    eax: 0,
+    ebx: 0,
+    ecx: 0,
+    edx: 0,
+};
+
+impl CpuIdWriter for CpuIdDump {
+    fn set_leaf(&mut self, leaf: u32, mut bits: Option<CpuIdResult>) {
+        // Many bits in 8000_0001h EDX, if present, mirror leaf 1h EDX. Maintain that before
+        // storing the bits.
+        if let Some(bits) = bits.as_mut() {
+            const MIRROR_MASK: u32 = 0b0000_0001_1000_0011_1111_0011_1111_1111;
+
+            if leaf == 0x0000_0001 {
+                // We're updating leaf 1h, so go fix up leaf 8000_0001h (if present)
+                match self.leaves.get_mut(&0x8000_0001) {
+                    Some(LeafOrSubleaves::Leaf(ext_info)) => {
+                        ext_info.edx &= !MIRROR_MASK;
+                        ext_info.edx |= bits.edx & MIRROR_MASK;
+                    }
+                    Some(_) => {
+                        panic!("extended feature information leaf (8000_0001h) had subleaves?");
+                    }
+                    None => {
+                        // No leaf 8000_0001h to mirror to (yet?)
+                    }
+                }
+            } else if leaf == 0x8000_0001 {
+                match self.leaves.get(&0x0000_0001) {
+                    Some(LeafOrSubleaves::Leaf(prior_bits)) => {
+                        bits.edx &= !MIRROR_MASK;
+                        bits.edx |= prior_bits.edx & MIRROR_MASK;
+                    }
+                    Some(_) => {
+                        panic!("feature information leaf (01h) had subleaves?");
+                    }
+                    None => {
+                        // No leaf 1h to shadow (yet?)
+                    }
+                }
+            };
+        }
+
+        if let Some(bits) = bits {
+            self.leaves.insert(leaf, LeafOrSubleaves::Leaf(bits));
+        } else {
+            self.leaves.remove(&leaf);
+        }
+
+        self.update_max_leaves();
+    }
+    fn set_subleaf(&mut self, leaf: u32, subleaf: u32, bits: Option<CpuIdResult>) {
+        if let Some(bits) = bits {
+            match self
+                .leaves
+                .entry(leaf)
+                .or_insert(LeafOrSubleaves::Subleaf(HashMap::new()))
+            {
+                LeafOrSubleaves::Leaf(_) => {
+                    panic!("adding a subleaf where there's a leaf. no");
+                }
+                LeafOrSubleaves::Subleaf(leaves) => {
+                    leaves.insert(subleaf, bits);
+                }
+            }
+        } else {
+            self.leaves.get_mut(&leaf).map(|ent| {
+                if let LeafOrSubleaves::Subleaf(leaves) = ent {
+                    leaves.remove(&subleaf);
+                } else {
+                    panic!("removing a subleaf when there's a leaf. no");
+                }
+            });
+        }
+
+        self.update_max_leaves();
+    }
+}
+
+impl CpuIdDump {
+    fn update_max_leaves(&mut self) {
```

Comment:
updating the max leaves on every write is naive, and this implementation is _really_ naive. If you want this change we could lean into "max leaf indicated by leaf 0, 4000_0000, or 8000_0000 is accurate" and check if they need changes on write.